ENH: Add annotations for `np.lib.arrayterator`

BvB93 opened the pull request:
This PR adds annotations for the `Arrayterator` class in `np.lib.arrayterator`.

Two things of note:
* The `__getattr__` method of `Arrayterator` wraps around the underlying `ndarray`s attributes.
  To simulate this behavior the former has been typed such that it subclasses the latter 
  (even though, strictly speaking, this isn't true at runtime).
* The `dtype` parameter was missing from the `Arrayterator.__array__` method. It has now been added.

Commit:
ENH: Add annotations for `np.lib.arrayterator`

Commit:
TST: Add typing tests for `np.lib.arrayterator`

rgommers commented on numpy/lib/arrayterator.py:
```diff
@@ -124,14 +124,17 @@ def __getitem__(self, index):
         return out
 
-    def __array__(self):
+    def __array__(self, dtype=None):
```
Looks to me like `dtype` should be a positional argument:
```
>>> x = np.arange(5)
>>> x.__array__(np.float32)
array([0., 1., 2., 3., 4.], dtype=float32)
>>> x.__array__(dtype=np.float32)
Traceback (most recent call last):
  File "<ipython-input-43-6adb0d04469e>", line 1, in <module>
    x.__array__(dtype=np.float32)
TypeError: __array__() takes no keyword arguments
```

BvB93 replied:
Based on some trial and error it seems that the `dtype` parameter must have a default value but can otherwise be either position-only or positional-or-keyword (`ndarray` being an example of the former).

Secondly, I'm not convinced anymore that this change is even necessary in the first place,
as I was initially under the (incorrect) impression that a `dtype` parameter would be required here 
if one wants to use it with the likes of, _e.g._, `np.array(x, dtype=<insert dtype different from x's dtype>`).

I think it might be best to revert this change as well and then make the `npt._SupportsArray` protocol a bit more flexible, such that it can accomodate `__array__` without the `dtype` parameter.

rgommers replied:
Yes, adding the `dtype` keyword is not needed, reverting sounds good.

rgommers commented on numpy/lib/arrayterator.py:
```diff
@@ -28,8 +28,8 @@ class Arrayterator:
-    var : array_like
-        The object to iterate over.
+    var : array
+        The array to iterate over.
```
This change looks incorrect. From a few lines above: _"`Arrayterator` can be used with any object that supports multidimensional slices."_

BvB93 replied:
That's a good point actually, I'll revert the change.

rgommers reviewed (changes requested):
Thanks @BvB93. Inheriting from `ndarray` in the stub seems reasonable. The changes to the class itself need a tweak.

Commit:
MAINT: Relax the signature of the `__array__` protocol

Implementations of `__array__` do not require the `dtype` parameter

Commit:
MAINT: Tighten the type-constraints of the `__array__` dtype parameter

The likes of `np.array(..., dtype=dtype)` will always pass `np.dtype` instances to `__array__(self, dtype=...)`, so that latter should not have to take arbitrary dtype-like objects but only actual `np.dtype` instances.

rgommers commented:
rebased to avoid the lint-check early CI termination.

rgommers reviewed (approved):
This LGTM now, so let's get it in. Thanks @BvB93

rgommers commented on numpy/typing/tests/data/reveal/flatiter.py:
```diff
@@ -14,4 +14,4 @@
 reveal_type(a[:])  # E: numpy.ndarray[Any, numpy.dtype[numpy.str_]]
 reveal_type(a.__array__())  # E: numpy.ndarray[Any, numpy.dtype[numpy.str_]]
-reveal_type(a.__array__(np.float64))  # E: numpy.ndarray[Any, numpy.dtype[Any]]
+reveal_type(a.__array__(np.dtype(np.float64)))  # E: numpy.ndarray[Any, numpy.dtype[{float64}]]
```
One follow-up question: `numpy.dtype[{float64}]` is a little verbose. How hard would it be to make `numpy.ndarray[Any, numpy.float64]` work?
